### ISSUE TITLE
UNIT and HELP must not be present

### ISSUE THREAD

**Issue by roidelapluie** [OpenObservability/OpenMetrics#255]
In the spec:

> MetricFamily
> A MetricFamily MAY have zero or more Metrics. A MetricFamily MUST have a name, HELP, TYPE, and UNIT metadata. Every Metric within a MetricFamily MUST have a unique LabelSet.

The first MUST is not correct.

In ABNF:

```
metricfamily = *metric-descriptor *metric

metric-descriptor = HASH SP type SP metricname SP metric-type LF
metric-descriptor =/ HASH SP help SP metricname SP escaped-string LF
metric-descriptor =/ HASH SP unit SP metricname SP *metricname-char LF
```

**Comment by RichiH** [comment]
From my Slack with Julien just now: I believe it should be

> A MetricFamily MUST have a name, and TYPE metadata. It SHOULD have HELP, and UNIT metadata.

This also aligns with what's later in the spec:
> If the unit is known it SHOULD be provided.
> The value of a UNIT or HELP line MAY be empty. This MUST be treated as if no metadata line for the MetricFamily existed.

**Comment by brian-brazil** [comment]
> From my Slack with Julien just now: I believe it should be

The current wording is what you argued for. The text is saying that the values must always be present, but HELP and UNIT can be empty.

**Comment by RichiH** [comment]
I can't remember; may well be.

The ABNF uses `=/`, so the whole line is optional as ABNF takes precedence.

As of today, I am not convinced if forcing empty lines is a positive.

**Comment by brian-brazil** [comment]
That section of the spec is about the data model, not the wire format.

We don't force empty lines.

**Comment by fstab** [comment]
Related:
> MetricFamilies of type Info MUST have an empty Unit string.

**Comment by brian-brazil** [comment]
Yes, that's correct as a unit is nonsensical for info. As before that line is in the data model section, not wire format.
